node/core/pvf/src/prepare/worker.rs

```diff
@@ -314,59 +307,61 @@ async fn recv_request(stream: &mut UnixStream) -> io::Result<(Vec<u8>, PathBuf,
 /// The entrypoint that the spawned prepare worker should start with. The `socket_path` specifies
 /// the path to the socket used to communicate with the host.
 pub fn worker_entrypoint(socket_path: &str) {
-	worker_event_loop("prepare", socket_path, |mut stream| async move {
+	worker_event_loop("prepare", socket_path, |stream| async move {
+		let mutex = Arc::new(Mutex::new((stream, false)));
 		loop {
-			let (code, dest, preparation_timeout) = recv_request(&mut stream).await?;
-
+			let (code, dest, preparation_timeout) = {
+				let mut lock = mutex.lock().await;
+				// Unset the lock flag. We set it when either thread finishes.
+				lock.1 = false;
+				recv_request(&mut lock.0).await?
+			};
 			gum::debug!(
 				target: LOG_TARGET,
 				worker_pid = %std::process::id(),
 				"worker: preparing artifact",
 			);
 
-			// Create a lock flag. We set it when either thread finishes.
-			let lock = Arc::new(AtomicBool::new(false));
 			let cpu_time_start = ProcessTime::now();
 
 			// Spawn a new thread that runs the CPU time monitor. Continuously wakes up from
 			// sleeping and then either sleeps for the remaining CPU time, or kills the process if
 			// we exceed the CPU timeout.
-			let (stream_2, cpu_time_start_2, preparation_timeout_2, lock_2) =
-				(stream.clone(), cpu_time_start, preparation_timeout, lock.clone());
+			let (mutex_2, cpu_time_start_2, preparation_timeout_2) =
+				(mutex.clone(), cpu_time_start, preparation_timeout);
 			let handle =
 				thread::Builder::new().name("CPU time monitor".into()).spawn(move || {
-					task::block_on(async {
+					let rt = Runtime::new().unwrap();
+					rt.block_on(async {
 						cpu_time_monitor_loop(
 							JobKind::Prepare,
-							stream_2,
+							mutex_2,
 							cpu_time_start_2,
 							preparation_timeout_2,
-							lock_2,
 						)
 						.await;
 					})
 				})?;
 
 			// Prepares the artifact in a separate thread.
-			let result = match prepare_artifact(&code).await {
+			let compilation_result = prepare_artifact(&code).await;
+			let cpu_time_elapsed = cpu_time_start.elapsed();
+
+			let mut lock = mutex.lock().await;
+			if lock.1 {
+				// Monitor thread detected timeout and the process should be terminated soon,
+				// nothing to do.
+				let _ = handle.join();
+				continue
+			}
+			lock.1 = true;
+
+			let result = match compilation_result {
 				Err(err) => {
 					// Serialized error will be written into the socket.
 					Err(err)
 				},
 				Ok(compiled_artifact) => {
-					let cpu_time_elapsed = cpu_time_start.elapsed();
-
-					let lock_result =
-						lock.compare_exchange(false, true, Ordering::Relaxed, Ordering::Relaxed);
```

Comment:
Noticed that we weren't setting the flag in the error case above. My bad. I fixed that in the refactor.